submissions/vovan-zt/TiniWorld2/index.js

```diff
@@ -0,0 +1,42 @@
+class WorldInhabitant {
+   constructor(species, name, gender, saying) {
+      this.species = species;
+      this.name = name;
+      this.gender = gender;
+      this.saying = saying;
+      this.prop = [species, name, gender, saying];
+   }
+
+   getMeaning() {
+      return this.prop.map(meaning => meaning);
+   }    
+}
+
+class Human extends WorldInhabitant { 
+   constructor(species, name, gender, saying, legs = 2, hands = 2, friends = '') {
+      super(species, name, gender, saying);
+      this.legs = legs;
+      this.hands = hands;
+      this.prop = [species, name, gender, saying, legs, hands, friends];
```

Comment:
Class shouldn't handle properties it doesn't own.
Imagine we need to add or remove a shared property.
Current design will require making changes to all classes.
Encapsulation helps to avoid this. Think of hierarchy of 20 classes.

Reply:
class WorldInhabitant {
  constructor(species, name, gender, saying) {
    this.species = species;
    this.name = name;
    this.gender = gender;
    this.saying = saying;
    this.prop = [species, name, gender, saying];
  }

  getMeaning() {
    return this.prop.map((meaning) => meaning).join("; ");
  }
}
class Human extends WorldInhabitant {
  constructor(name, gender, saying, friends = 'no friends') {
    super("human", name, gender, saying);
    this.legs = 2;
    this.hands = 2;
    this.friends = friends;
    this.prop = [...this.prop, this.legs, this.hands, this.friends];
  }
}

If this approach is bad practice, then I will change it.


  getMeaning() {
    return ['species', 'name', 'gender', 'saying'].map(property => this[property]).join('; ');
  }

  getMeaning() {
    return `${super.getMeaning()}; ${this.legs}; ${this.hands}; ${this.friends?this.friends:'no friends'}`;
 };

Reply:
(2) 
```
getMeaning() {
return `${super.getMeaning()}; ${this.legs}; ${this.hands}; ${this.friends?this.friends:'no friends'}`};
};
```

This is a good approach for a child class. Overloaded method calls inherited version (so that parent class manages all its own properties) and adds properties that are owned by the child class. Very OOP compliant.

Also note, that array literal may comprise whatever legal expressions we want.
Consider the following fragment:
```
[
  'some text',
  localVariable,
  this.property,
  makeSomeText(),
].join("; ")

```